### DISCUSSION TITLE
1.1.8 - Corepack

### DISCUSSION THREAD

**Discussion by coreybutler** [coreybutler/nvm-windows#664] [General]
# NVM for Windows v1.1.8 Notes

Truth be told, I did not want to release this version right now. I am working on the successor project [rt](https://github.com/coreybutler/nvm-windows/discussions/565). v1.1.7 was supposed to be the last version before the rt release. However, notable changes in the Node ecosystem warrant a new version now.

**Node Corepack**

In September 2021, Node.js v16.9.0 introduced [corepack](https://nodejs.org/api/corepack.html). This experimental new feature allows transparent use of npm, pnpm, or yarn. To support this feature, NVM4W must download and process a different distribution file than it has used previously.

As a result, NVM for Windows 1.1.8 is being released to support corepack.

## How to Upgrade

All files can be found on the [release page](https://github.com/coreybutler/nvm-windows/releases/tag/1.1.8).

**If you do not have NVM for Windows installed**
Use the instructions in the README.

**If you already have an older version of NVM for Windows installed**
Download *nvm-update.zip* and extract it. You will find a file called *nvm-update.exe*. Double click to run it. This will prompt for the version you wish to upgrade to (pre-populated w/ 1.1.8).

nvm-update.exe can also be run from the command line using the `/S` flag to suppress the prompt. To do this, run `nvm-update.exe 1.1.8 /S`.

## What's New

Several long-requested features/fixes have been merged since v1.1.7. I did not get to test all of them as extensively as I wanted to, but I believe they are stable. These features include:

1. corepack support (see above)
2. Support for spaces in filepaths [#355](https://github.com/coreybutler/nvm-windows/pull/355).
3. `nvm install latest` - installs the latest patch version of Node (instead of latest minor).
4. `nvm install lts` - installs the latest LTS patch version (new).
5. `nvm use latest`, `nvm use lts`, and `nvm use newest` (new) now supported.
6. Elevated permissions applied only when necessary [#511](https://github.com/coreybutler/nvm-windows/pull/511).
7. `nvm current` (new) displays the active version.

Numerous edge case fixes have been applied as well.

## What didn't make it in?

1. Aliases - these were originally scheduled for 1.1.8, but I never got to work on them.

## WARNING

```
------- UPDATE NOTICE -------
Thanks to a kind sponsor, an extended validation code signing certificate is was acquired. Version 1.1.9 is code signed and should be used instead of v1.1.8.  
```

Version 1.1.8 is **not code signed**. Code-signed applications are trusted by Windows and will install easily. Users are prompted with a screen like this one:

![code signed](https://www.globalsign.com/application/files/9415/8088/1150/ms-authenticode-warning2.png)

Non-code-signed applications can _usually_ be installed, but they will throw a warning like this:

![non-code signing](https://sectigostore.com/static/images/code-sign-warning.png)

**Why?**

I paid for the last code signing certificates out of pocket, but they've expired. Sponsorships, to date, haven't even come close to covering the cost of a new code signing certificate (let alone the last one).

**Who will this affect?**

This won't affect most users, but enterprise users and those distributing NVM4W in corporate environments may run into problems.

**If your organization is impacted by this, I will accept sponsorship money to expedite the delivery of a code-signed version.** The required EV certificate is about $1000. Please direct message me on Twitter [@goldglovecb](https://twitter.com/goldglovecb) if you'd like to help cover this cost.

<hr /><em>This discussion was created from the release <a href='https://github.com/coreybutler/nvm-windows/releases/tag/1.1.8'>1.1.8 - Corepack</a>.</em>

**Comment by ceciliabarudi** [comment]
![image](https://user-images.githubusercontent.com/26797715/133464887-5fe45b94-4c7d-4463-8f3e-263a69be2bca.png)

Hey, a heads-up/question. I get this after downloading the nvm-update.zip , is it normal? (sorry if this isn't the adequate place for this comment)

**Reply by coreybutler** [reply]
Unsigned Go apps sometimes throw a false positive, suggesting a trojan exists when it doesn't. I wouldn't call it "normal", but it's also not surprising... especially if your antivirus definitions are a little outdated.

Which antivirus is identifying this as a threat?

**Reply by loriswit** [reply]
I'm getting the same alert. My only antivirus is the default Windows Defender, running on Windows 10 21H1.

**Reply by coreybutler** [reply]
I cannot recreate the issue in Windows Defender, but the Microsoft support suggests that the `!ml` identifier in the virus definition relies on machine learning... in other words, Windows Defender is incorrectly guessing there is an issue.

**Reply by ceciliabarudi** [reply]
Yes, I got the warning with Windows Defender as well. Thank you for the explanation, I've given it a ⬆️ so people who stumble with this see it first. I will try installing again today.

**Reply by zipzit** [reply]
Hmm.  Not only did I get an immediate virus warning, but my Windows OS blocked the download. False positive or not, the nvm-update.zip version simply wouldn’t download.

**Reply by coreybutler** [reply]
I have added a [note in the wiki](https://github.com/coreybutler/nvm-windows/wiki/Common-Issues#antivirus) with basic help for Windows Defender users.

**Comment by kccarter76** [comment]
our endpoint security actually quarantined the nvm.exe file. you really need to scan for virus. the maintenance release before this one passes the scan

**Reply by coreybutler** [reply]
There is no virus, and I do run a quick check... but there are over 100 antivirus vendors. Furthermore, this happens for new unsigned apps. If someone wants to donate the $1000+ for the required code signing certificate, then I'd be happy to make a signed release.

**Comment by kccarter76** [comment]
![image](https://user-images.githubusercontent.com/5841402/133511713-8771b4d9-6afe-4621-b6ef-9ceb53df867f.png)

**Comment by coreybutler** [comment]
Folks - yes, the antivirus tools may mark unrecognized executables as a virus. This happened with the initial 1.1.7 release as well. 

Notice how most of the threats are labeled with "ML" or "!ml" - that stands for machine learning with most antivirus providers. In other words, the antivirus suites are guessing... these are not signature threats. A signature threat is one which has been confirmed (i.e. the signature of the executable has been explicitly identified as containing a virus).

The antivirus suites often use ML patterns on "untrusted" executables. An untrusted executable is one which is not code-signed. Apps signed with basic code signing certificates are trusted by Windows after enough people install the app. Apps signed with an extended validation certificate are immediately trusted by Windows. This version of NVM4W is not code-signed at all, because my last certificate expired in 2020. The extended validation certificates cost around $1000, which is more than the lifetime earnings of this project. So, unless someone wants to cover that cost, I have no plans to code sign this release (as noted in the release notes).

The best way to help is to register a "false threat" with your antivirus provider. I cannot test all the different antivirus software. Most antivirus software will let you make an exception for apps, so I recommend doing that if you run into quarantines/deletions.

**Reply by kccarter76** [reply]
does anyone know what has happened to all the node releases, trying to install 14.15.5 but the release tag is no longer available on github.

![image](https://user-images.githubusercontent.com/5841402/133514424-3503bb09-8297-41e8-a7ef-6801a99d9384.png)

**Reply by kccarter76** [reply]
after undoing the quarantine and installing the updated software the above issue with not find the node to download is resolved.

**Comment by davidgeary** [comment]
Can anyone confirm the checksum for _nvm-update.zip_? The accompanying text file states that it should be `8c5739317a14230aa1f8ca8070d98a36` but I get the following results using 7-Zip's CRC-SHA utility:

``` sh
Name: nvm-update.zip
Size: 3420632 bytes (3340 KiB)
CRC32: FBC0BCF6
CRC64: 078BFD3D4C73E774
SHA256: 749AF837AD391858D39FA5FD0DFCF40D7F347CA2D233C70ED8871A6EBA081283
SHA1: BBC90F60EEE0D87AD580B622C210F68541447B1C
BLAKE2sp: 231AB60EACFEE0F3B1FFE378B092EC5D5B1889785B39C530AE1EDC9AC4E0E032
```

Or has it been calculated using some other algorithm perhaps?

**Reply by coreybutler** [reply]
The checksums use the MD5 hash as generated by certutil. See https://github.com/coreybutler/nvm-windows/blob/master/build.bat#L59

**Reply by davidgeary** [reply]
@coreybutler Ah, of course - duh!  So used to MD5 being considered obsolete in crypto that I completely forgot about it for checksums.  Thanks for the quick response.

**Reply by ConPac** [reply]
Maybe not the place to tell this, but please consider using in build.bat a non-obsolete hash algorithm supported by certutil, like SHA256, and include its identifier in the checksum filename, like nvm-update.zip.SHA256-checksum.txt.
And many thanks for your nice work.

**Comment by ghost** [comment]
The Linux Foundation has announced a project for producing free code-signing certificates. It is still in beta, but will be available in the near future. https://www.sigstore.dev/

**Reply by coreybutler** [reply]
Wow - I was unaware of this. I'll reach out to them. Thanks for bringing my attention to this.

**Reply by ghost** [reply]
Look into the Fulcio sub-project (https://docs.sigstore.dev/fulcio/overview). It is the free Certificate Authority. It is a bit of a change from the normal paradigm. The certificates are only good for 20 mins. I think the idea might be that your release pipeline automatically generates a single-use certificate, and uses it to sign your release files. I guess we'll see when it's ready.

**Reply by coreybutler** [reply]
The bigger challenge at this point is whether the CA is trusted by the operating system. Microsoft SmartScreen maintains a strict trust list, which usually requires a partnership with Microsoft. Microsoft was the one company I _didn't_ see in the Sigstore list. As of right now, it looks like this service will help confirm libraries are trustworthy (similar to checksums), but it doesn't mean apps signed with these certs will be trusted by operating systems just yet. It would be great if it evolved to this point though. I'll be keeping an eye on this effort, but it looks like an EV certificate is still required for immediate trust with Microsoft (for now).

**Reply by ghost** [reply]
This has been exactly our challenge with Microsoft. Costs thousands of dollars so that we can be shipped a physical USB key we have to use when code signing. Ridiculous.

**Reply by coreybutler** [reply]
I will make a more formal announcement later, but someone stepped up to sponsor the code signing certificate. I am currently going through the validation process with the certificate authority. I've done what I can to expedite, but it may still be a couple of weeks before the certificate is issued/arrives. When it does, I will release code signed v1.1.8 executables.

**Reply by ghost** [reply]
That's great news! Thanks for your hard work. If we developed more consistently on Windows I would definitely be reaching out.

**Reply by coreybutler** [reply]
@jchambliss keep an eye on things towards the end of the year. The new rt project (NVM4W successor) is cross-platform and cross-runtime.

**Comment by longsoft-dev** [comment]
I've got a work-around. Just download and install the previous release - 1.1.7. Then download the 1.1.8 zip version. Unzip and copy the nvm.exe to <nvm 1.1.7 home dir> and overwrite the old one ;-)

**Reply by coreybutler** [reply]
This is exactly what the updater does, except it copies over all of the files you need, not just nvm.exe.

**Comment by Taweesakpintakam7m1** [comment]
> # NVM for Windows v1.1.8 Notes
> Truth be told, I did not want to release this version right now. I am working on the successor project [rt](https://github.com/coreybutler/nvm-windows/discussions/565), which was supposed to be the last version before the rt release. However, notable changes in the Node ecosystem warrant a new version now.
> 
> **Node Corepack**
> 
> In September 2021, Node.js v16.9.0 introduced [corepack](https://nodejs.org/api/corepack.html). This experimental new feature allows transparent use of npm, pnpm, or yarn. To support this feature, NVM4W must download and process a different distribution file than it has used previously.
> 
> As a result, NVM for Windows 1.1.8 is being released to support corepack.
> 
> ## How to Upgrade
> All files can be found on the [release page](https://github.com/coreybutler/nvm-windows/releases/tag/1.1.8).
> 
> **If you do not have NVM for Windows installed** Use the instructions in the README.
> 
> **If you already have an older version of NVM for Windows installed** Download _nvm-update.zip_ and extract it. You will find a file called _nvm-update.exe_. Double click to run it. This will prompt for the version you wish to upgrade to (pre-populated w/ 1.1.8).
> 
> nvm-update.exe can also be run from the command line using the `/S` flag to suppress the prompt. To do this, run `nvm-update.exe 1.1.8 /S`.
> 
> ## What's New
> Several long-requested features/fixes have been merged since v1.1.7. I did not get to test all of them as extensively as I wanted to, but I believe they are stable. These features include:
> 
> 1. corepack support (see above)
> 2. Support for spaces in filepaths [#355](https://github.com/coreybutler/nvm-windows/pull/355).
> 3. `nvm install latest` - installs the latest patch version of Node (instead of latest minor).
> 4. `nvm install lts` - installs the latest LTS patch version (new).
> 5. `nvm use latest`, `nvm use lts`, and `nvm use newest` (new) now supported.
> 6. Elevated permissions applied only when necessary [#511](https://github.com/coreybutler/nvm-windows/pull/511).
> 7. `nvm current` (new) displays the active version.
> 
> Numerous edge case fixes have been applied as well.
> 
> ## What didn't make it in?
> 1. Aliases - these were originally scheduled for 1.1.8, but I never got to work on them.
> 
> ## WARNING
> Version 1.1.8 is **not code signed**. Code-signed applications are trusted by Windows and will install easily. Users are prompted with a screen like this one:
> 
> ![code signed](https://camo.githubusercontent.com/649275add784848014ad51fe1a5b0b9130eb50435f4193a946a20b3d79907905/68747470733a2f2f7777772e676c6f62616c7369676e2e636f6d2f6170706c69636174696f6e2f66696c65732f393431352f383038382f313135302f6d732d61757468656e7469636f64652d7761726e696e67322e706e67)
> 
> Non-code-signed applications can _usually_ be installed, but they will throw a warning like this:
> 
> ![non-code signing](https://camo.githubusercontent.com/e99ddc9fe9848033b6f25d1459c6e2acb9a028b582da6dac03812dd9c05011ed/68747470733a2f2f7365637469676f73746f72652e636f6d2f7374617469632f696d616765732f636f64652d7369676e2d7761726e696e672e706e67)
> 
> **Why?**
> 
> I paid for the last code signing certificates out of pocket, but they've expired. Sponsorships, to date, haven't even come close to covering the cost of a new code signing certificate (let alone the last one).
> 
> **Who will this affect?**
> 
> This won't affect most users, but enterprise users and those distributing NVM4W in corporate environments may run into problems.
> 
> **If your organization is impacted by this, I will accept sponsorship money to expedite the delivery of a code-signed version.** The required EV certificate is about $1000. Please direct message me on Twitter [@goldglovecb](https://twitter.com/goldglovecb) if you'd like to help cover this cost.
> 
> _This discussion was created from the release [1.1.8 - Corepack](https://github.com/coreybutler/nvm-windows/releases/tag/1.1.8)._
